### PR TITLE
MapJoin support, test fixes for hive-0.11

### PR DESCRIPTION
**Pull request by YuryYudin**

### CODE REVIEW

**Comment by harveyfeng** [comment]
Awesome! How much does the Hive unit test coverage change with these patches (should we update https://github.com/amplab/shark/blob/master/src/test/tests_pass.txt)?

**Comment by YuryYudin** [comment]
This is in progress right now - will submit another PR for tests_pass and
some more test fixes in nearest couple working days.

Thanks,
Yury

**Comment by harveyfeng** [comment]
Hi Yury, just wondering how work on this patch is going?

Thanks!

**Comment by YuryYudin** [comment]
Hi Harvey,

Will submit a PR tomorrow, sorry for delay.

Thanks,
Yury

File: src/main/scala/shark/execution/CommonJoinOperator.scala [hunk 1]
```diff
@@ -204,10 +204,12 @@ object CommonJoinOperator {
   val LEFT_SEMI_JOIN = JoinDesc.LEFT_SEMI_JOIN
 
   /**
-   * Handles join filters in Hive. It is kind of buggy and not used at the moment.
+   * Handles join filters in Hive
    */
   def isFiltered(row: Any, filters: JavaList[ExprNodeEvaluator], ois: JavaList[ObjectInspector])
   : Boolean = {
+    if(row == null) return true
```
Comment (harveyfeng) [inline src/main/scala/shark/execution/CommonJoinOperator.scala:211]: Space after `if` here. Should change other instances in this PR too.

Comment (harveyfeng) [inline src/main/scala/shark/execution/CommonJoinOperator.scala:211]: Add a short comment for why the default return value is `true` for `null` rows. Would it work if the `null` check is moved to
`private def generate(previousRightFiltered: Boolean, previousRightTable: Array[Object])` in JoinOperator? That would fit with the "discard if true" description in there.

File: src/main/scala/shark/execution/JoinOperator.scala [hunk 2]
```diff
@@ -19,6 +19,7 @@ package shark.execution
 
 import java.util.{HashMap => JHashMap, List => JList}
 
+import scala.collection.mutable.Queue
 import scala.collection.mutable.ArrayBuffer
```
Comment (harveyfeng) [inline src/main/scala/shark/execution/JoinOperator.scala:23]: alphabetical ordering for classes in the same package

File: src/main/scala/shark/execution/JoinOperator.scala [hunk 3]
```diff
@@ -139,47 +141,229 @@ class JoinOperator extends CommonJoinOperator[JoinDesc, HiveJoinOperator]
...
+        processNext()
+      }
+
+      outputs.dequeue
+    }
+    
+    private def processNext() {
+      if(!done) {
+        // if not set as finished, iterate the next
+        var continue = (outputs.isEmpty) // if not element in the queue, the try to get
```
Comment (harveyfeng) [inline src/main/scala/shark/execution/JoinOperator.scala:178]: the => then

File: src/main/scala/shark/execution/JoinOperator.scala [hunk 4]
```diff
@@ -139,47 +141,229 @@ class JoinOperator extends CommonJoinOperator[JoinDesc, HiveJoinOperator]
...
       }
+      
+      // set the most right table columns
+      if(!leftFiltered) {
+        setColumnValues(leftTable, row, index.toByte, offsets)
+      }
+      
+      entireLeftFiltered = entireLeftFiltered && leftFiltered
+      
+      if(entireLeftFiltered)
```
Comment (harveyfeng) [inline src/main/scala/shark/execution/JoinOperator.scala:338]: Add braces for multi-lined `if ... else`

**Comment by harveyfeng** [comment]
Going to take a closer look at this later today. Linking https://github.com/amplab/shark/pull/161 as a reference for now.

Btw, where a few of the included commits (e.g., 6fc3d3f) cherry-picked and then modified during a merge? Not a blocker of course, but I think maintaining a clean commit history will definitely pay dividends moving forwards.

**Comment by YuryYudin** [comment]
Hi Harvey,

Your guess is right, we cherry-picked that from the #161 3 month ago. We
needed a fix for the issue which wasn't yet integrated into the amplab's
hive0.11, so went ahead and integrated it into our branch and made couple
fixes on top of it. We haven't modified it during a merge, as it should
have been applied cleanly. Was it modified afterwards?

BTW, the comments you provided are all within that commit, so most likely
have already been addressed by Cheng Hao. We can rebase over a new one if
it's needed, as long as this is almost the first commit in our PR it should
go cleanly.

Thanks,
Yury

**Comment by harveyfeng** [comment]
Looks like the commit was amended since the cherry-pick - it includes style changes and clearer documentation. One path is to just merge this PR into the hive0.11 branch first, update branch-0.9, which is a working branch for the hive0.11-onto-master merge, and then cherry-pick (+resolve conflicts) in #161 into branch-0.9.

**Comment by harveyfeng** [comment]
Thanks!

Cherry-picked these into master and branch-0.9.

**Comment by harveyfeng** [comment]
FYI  going to delete hive0.11 branch soon now that it's been merged into master

**Pull request closed without merging**